adoc/deployment-ecp.adoc

```diff
@@ -44,11 +48,21 @@ include::deployment-terraform-example.adoc[tags=tf_openstack]
 +
 [TIP]
 ====
-You can set timezone in before deploying the nodes by modifying the files:
+You can set timezone in before deploying the nodes by modifying the file:
 
-* `~/my-deployment/cloud-init/master.tpl`
-* `~/my-deployment/cloud-init/worker.tpl`
+* `~/my-deployment/cloud-init/common.tpl`
 ====
+. Optional: If you absolutely need to be able to ssh into your cluster nodes using password instead of key-based authentication, this is the best time to do it massively for all of your nodes. Alternatively you can do it also later, but manually. To do that you need to modify the cloud-init configuration and comment-out the related SSH configuration:
+* `~/my-deployment/cloud-init/common.tpl`
++
+----
+# Workaround for bsc#1138557 . Disable root and password SSH login
```

Comment:
I am a bit confused here - why do we hack ssh_config to disable password authentication when sles/root users don't have any password? And then tell customers they can hack it back if they want to set passwords.

I would expect that password authentication is off just by having no password set for system users.
I don't understand what is this feature good for.

Also vmware & bare-metal don't have this config hacks, so why openstack?

Reply:
First off let me very quickly clarify that is what the team decided when they answered to the question _how would you prefer to be able to ssh to the caasp nodes?_ 

**7 votes** (including the PM): Only with SSH Key based authention. Password-logins should be disabled.
0 votes: Only with password logins.
1 votes: Both.

Now, going to your questions:
> I would expect that password authentication is off just by having no password set for system users.

It is considered best security practice to disable the password authentication for systems running in production. The customer might set a password for either `root` or `sles` but this should not be related to _ssh-ing_ into the cluster remotely. If they want to allow ssh password logins they have to explicitly do it by accepting the security consequences of their actions. If you have password authentication enabled on your SSH service and your root password is weak, chances you are going to get hacked are pretty high. That's why disabling root login is one of the oldest and most used techniques to avoid system compromise on fresh OS installations. You can check this out by leaving the SSH installation by default,parse the `/var/log/secure` file and see how many brute force attacks you will receive within a few hours -- you will be surprised. Even if you are using strong passwords, it's always recommended to disable root login and use a primary SSH user instead, so then if you need to gain root access you can do it using `su` or `sudo` command if required.

It is true that `sles` and `root` users have empty passwords. And this can be a pretty bad thing if you want to keep attackers out of your SSH servers. That's why the best thing you can do is to disable remote logins for accounts with an empty password.

> I don't understand what is this feature good for.

It's hardening the security by mitigating SSH based attacks.

> Also vmware & bare-metal don't have this config hacks, so why openstack?

I guess we have to do it there as well.

Reply:
> 
> 
> First off let me very quickly clarify that is what the team decided when they answered to the question _how would you prefer to be able to ssh to the caasp nodes?_
> 
> **7 votes** (including the PM): Only with SSH Key based authention. Password-logins should be disabled.
> 0 votes: Only with password logins.
> 1 votes: Both.
> 

I have to give a bit feedback as I am not sure how all of this matches to real world ;-).

Did anyone discuss this with customers and what their setup looks like?

So far - in case SSH key based logins are used - the majority of customers to not password protect the private keys that are stored on the servers / jump hosts.
Even up to SES 4 we explained how to set this up and with Salt we also do not have password protected private keys ;-).
AFAIK als SLE-HA also uses non-password protected SSH keys.

> Now, going to your questions:
> 
> > I would expect that password authentication is off just by having no password set for system users.
> 
> It is considered best security practice to disable the password authentication for systems running in production. 
> The customer might set a password for either `root` or `sles` but this should not be related to _ssh-ing_ into the cluster remotely.
> If they want to allow ssh password logins they have to explicitly do it by accepting the security consequences of their actions. 
> If you have password authentication enabled on your SSH service and your root password is weak, chances you are going to get hacked are pretty high. 
> That's why disabling root login is one of the oldest and most used techniques to avoid system compromise on fresh OS installations.
> You can check this out by leaving the SSH installation by default,parse the `/var/log/secure` file and see how many brute force attacks you will receive within a few hours -- you will be surprised. Even if you are using strong passwords, it's always recommended to disable root login and use a primary SSH user instead, so then if you need to gain root access you can do it using `su` or `sudo` command if required.

While all of this is true and should be recommended - the majority of customers do have personalized authentication setup on the servers using i.e. pam with LDAP or AD connection.
Admins login as their user with their LDAP / AD password and then do sudo / ssh to other systems using keys (most often not password protected) or again username + password.


> 
> It is true that `sles` and `root` users have empty passwords. And this can be a pretty bad thing if you want to keep attackers out of your SSH servers. That's why the best thing you can do is to disable remote logins for accounts with an empty password.
> 
> > I don't understand what is this feature good for.
> 
> It's hardening the security by mitigating SSH based attacks.
> 
> > Also vmware & bare-metal don't have this config hacks, so why openstack?
> 
> I guess we have to do it there as well.

IMO we have to support "hardening" and we have to deliver a "hardening guide" similar to what we deliver for SLES - but we also should have defaults and base our solution on what matches to the customers current implementations and security behaviours.

So yes - password protected ssh keys with SSH agent must be supported using non-root users and sudo - but on the other hand also non-password protected and available ssh keys should be used automatically without the need to run ssh-agent and ssh-add for non-password protected private ssh keys.

just my 2 cents ;-)

Reply:
> 
> 
> First off let me very quickly clarify that is what the team decided when they answered to the question _how would you prefer to be able to ssh to the caasp nodes?_
> 
> **7 votes** (including the PM): Only with SSH Key based authention. Password-logins should be disabled.
> 0 votes: Only with password logins.
> 1 votes: Both.
> 

I have to give a bit feedback as I am not sure how all of this matches to real world ;-).

Did anyone discuss this with customers and what their setup looks like?

So far - in case SSH key based logins are used - the majority of customers to not password protect the private keys that are stored on the servers / jump hosts.
Even up to SES 4 we explained how to set this up and with Salt we also do not have password protected private keys ;-).
AFAIK als SLE-HA also uses non-password protected SSH keys.

> Now, going to your questions:
> 
> > I would expect that password authentication is off just by having no password set for system users.
> 
> It is considered best security practice to disable the password authentication for systems running in production. 
> The customer might set a password for either `root` or `sles` but this should not be related to _ssh-ing_ into the cluster remotely.
> If they want to allow ssh password logins they have to explicitly do it by accepting the security consequences of their actions. 
> If you have password authentication enabled on your SSH service and your root password is weak, chances you are going to get hacked are pretty high. 
> That's why disabling root login is one of the oldest and most used techniques to avoid system compromise on fresh OS installations.
> You can check this out by leaving the SSH installation by default,parse the `/var/log/secure` file and see how many brute force attacks you will receive within a few hours -- you will be surprised. Even if you are using strong passwords, it's always recommended to disable root login and use a primary SSH user instead, so then if you need to gain root access you can do it using `su` or `sudo` command if required.

While all of this is true and should be recommended - the majority of customers do have personalized authentication setup on the servers using i.e. pam with LDAP or AD connection.
Admins login as their user with their LDAP / AD password and then do sudo / ssh to other systems using keys (most often not password protected) or again username + password.


> 
> It is true that `sles` and `root` users have empty passwords. And this can be a pretty bad thing if you want to keep attackers out of your SSH servers. That's why the best thing you can do is to disable remote logins for accounts with an empty password.
> 
> > I don't understand what is this feature good for.
> 
> It's hardening the security by mitigating SSH based attacks.
> 
> > Also vmware & bare-metal don't have this config hacks, so why openstack?
> 
> I guess we have to do it there as well.

IMO we have to support "hardening" and we have to deliver a "hardening guide" similar to what we deliver for SLES - but we also should have defaults and base our solution on what matches to the customers current implementations and security behaviours.

So yes - password protected ssh keys with SSH agent must be supported using non-root users and sudo - but on the other hand also non-password protected and available ssh keys should be used automatically without the need to run ssh-agent and ssh-add for non-password protected private ssh keys.

just my 2 cents ;-)

Reply:
> So yes - password protected ssh keys with SSH agent must be supported using non-root users and sudo

Great :)


> but on the other hand also non-password protected and available ssh keys should be used automatically without the need to run ssh-agent and ssh-add for non-password protected private ssh keys.

Exactly, you're right. We are still in Beta3 and this feature was not part of the MVP. The idea is that the customer who doesn't want to use `ssh-agent` they should be able to provide the path to their private ssh key as well. While this is easily done in Terraform, we need to change our ssh functionality in skuba. I would like to ask you to open a feature request so we priortize that into the roadmap.

And of course, many thanks for your feedback, it's very welcome to have customer's insights :)